File: Leetcode/3.无重复字符的最长子串.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> hashset;
        int res = 0;
        int right = 1;
        int n = s.size();

        for (int i = 0; i < n; i ++ ) {
            if (i > 0) {
                hashset.erase(s[i - 1]);
            } else {
                hashset.insert(s[i]);
            }

            while (right < n && !hashset.count(s[right])) {
                hashset.insert(s[right]);
                right ++ ;
            }

            res = max(res, right - i);
        }

        return res;
    }
};

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size(), maxn = 0, point1 = 0, point2 = 1;
        if (n < 2) {
            return n;
        } 

        while (point2 < n) {
            for (int i = point1; i < point2; ++ i) {
                if (s[i] == s[point2]) {
                    point1 = i + 1;
                    break;
                }
            }
            maxn = max(point2 - point1 + 1, maxn);
            ++ point2;
        }

        return maxn;
    }
};

File: Leetcode/3.无重复字符的最长子串.go
func max(i int, j int) int {
	if i <= j {
		return j
	}
	return i
}

func lengthOfLongestSubstring(s string) int {
	if len(s) < 2 {
		return len(s)
	}
	map_ := make(map[byte]int)
	j := 1
	res := 1
	for i := 0; i < len(s); i++ {
		if i == 0 {
			map_[s[i]] = 1
		} else {
			delete(map_, s[i-1])
		}
		for j < len(s) && map_[s[j]] == 0 { //  map中没有的元素，默认是零值
			map_[s[j]] = 1
			j++
		}
		res = max(res, j-i)
	}
	return res
}